### DISCUSSION TITLE
mDNS support on Pico

### DISCUSSION THREAD

**Discussion by ondrej1024** [micropython/micropython#10932] [RP2040 / Pico] [answered]
Hi,

is there mDNS support enabled on the RP2040/Pico port of Micropython (like ESP32 port)?
If it is, how can it be configured?

I am running this MP version:
`MicroPython v1.19.1-910-g4937174b4 on 2023-03-01; W5500-EVB-Pico with RP2040`

Network connection is active but I cannot see the mDNS info on the network.

**Comment by ondrej1024** [comment]
So I dug further into this issue and found some [documentation about the mDNS service implemented in the lwip stack](https://github.com/lwip-tcpip/lwip/blob/6ca936f6b588cee702c638eee75c2436e6cf75de/doc/mdns.txt).

This states to call this function to initialise the service:
`mdns_resp_init()`

and then start responding on a netif by calling this function:
`mdns_resp_add_netif()`

On the Pico port, the first function is actually called in [main.c](https://github.com/micropython/micropython/blob/master/ports/rp2/main.c) but the second function never seems to be called. This leaves the service inactive. Looks like a bug in the MP Pico port. 

So I wonder which part of the MP implementation should call `mdns_resp_add_netif()`? It has to be done probably after the network interface has been created by the driver. 

Anyone who can shed some light on this?

**Comment by ondrej1024** [comment]
I think [this bug report](https://github.com/micropython/micropython/issues/8946) answers my question.

**Comment by vinux84** [comment]
Hi, Just curious, were you able to get mDNS to work on the RP2040?

**Reply by ondrej1024** [reply]
I ended up using the [micropython-mdns](https://github.com/cbrand/micropython-mdns) package which works great.

**Comment by scruss** [comment]
(this may not be a useful answer)

I can confirm local DNS (not mDNS) with the nightly build *MicroPython v1.23.0-preview.324.gd11ca092f on 2024-04-17; Raspberry Pi Pico W with RP2040*

Set the hostname **before** you initialize the network:
```python3
    import network
    network.hostname("picowX")
    ...
```
For some reason, my LAN seems to prefer *.ht.home* over *.local*, but I can do this:
```
ping picowX.ht.home
PING picowX.ht.home (192.168.2.138) 56(84) bytes of data.
64 bytes from picowX.ht.home (192.168.2.138): icmp_seq=1 ttl=255 time=293 ms
64 bytes from picowX.ht.home (192.168.2.138): icmp_seq=2 ttl=255 time=123 ms
64 bytes from picowX.ht.home (192.168.2.138): icmp_seq=3 ttl=255 time=145 ms
...
```

**Reply by cwalther** [reply]
> For some reason, my LAN seems to prefer _.ht.home_ over _.local_, but I can do this:
> 
> ```
> ping picowX.ht.home
> PING picowX.ht.home (192.168.2.138) 56(84) bytes of data.
> ```

That does not look like mDNS to me. That looks like your LAN router registers the name `picowX.ht.home` in its own DNS server (using information it has from also being the DHCP server). I suspect it’s the router answering the query here, not the Pico.

(I can’t try it myself because I don’t have a Pico W.)

**Reply by scruss** [reply]
ah, right — that would be it.

**Reply by ShadowLNC** [reply]
I couldn't figure out why I could `ping pico` on Linux (WSL, Rasperry Pi), but not on Windows (Powershell, CMD) - I suspect that my router is doing the same thing here (I can see it responds to `dig`).

It also seems that Linux will resolve TLDs as hostnames if a DNS record exists, but that Windows will not, although `ping pico.` (note the dot) will resolve, so for some reason Windows will only accept a DNS TLD if it ends with a dot (which is usually implicit), otherwise it assumes it must be mDNS or NetBIOS.

**Comment by vinux84** [comment]
I have been able to get the same package to work that @ondrej1024 uses, just bare bones, pulling the example out of that repository. 

But has anyone ever got this to work with python http server, or any high level working web server?

Basically Im wanting a user to be able to access pico w with :  hostname.local and be redirected to a home web page, hosted from the pico. This isn't my strong suit by any means and I could be missing something very obvious, but could not get it to work.

**Reply by mendenm** [reply]
@vinux84 I think it should work with microdot, which is itself a really nice web server for the Pico.

**Reply by osarhan** [reply]
Not sure if this helps.. but I hard coded the hostname in the micropython files.

The default is `PicoW` on the board.  (try to access it using `picow.local` or `picow.lan`)

Here's what I did to bake it in: https://github.com/micropython/micropython/issues/17976#issuecomment-3246504256

**Comment by ondrej1024** [comment]
The native MP (lwIP) mDNS support still does not seem to be fixed in MP 1.26 for Pico W. 
Can anyone confirm that ? Or is it just me doing something wrong?

**Reply by scruss** [reply]
looks like [LWIP_MDNS_RESPONDER](https://github.com/micropython/micropython/blob/27544a2d81da5b0d804a932d98d680f121a22b8f/ports/rp2/main.c#L154) is never defined in any config file, so mDNS is likely never started

**Reply by ondrej1024** [reply]
The thing is that in the release note of MP 1.25 it states:

The **rp2 port** sees the introduction of many new RP2350 boards, including the
    Pico 2 W, as well as support for PSRAM with size autodetection.  The PIO
    interface now supports `side_pindir` selection, and SPI allows the MISO pin
    to be unspecified.  Both the I2C and SPI classes now have the bus
    identifier as an optional argument with a default based on the board
    configuration. WPA3 is now supported on the Pico W and Pico 2 W in both AP
    and STA modes. Lost WiFi events due to code executing on the second core
    have now been fixed, **mDNS has been fixed**, and `rp2.bootsel_button()` and
    USB sleep now work on RP2350.  ROMFS support has been added but is not
    enabled on any board by default

```
rp2 port:
   ...
    - cyw43_configport: fix cyw43 mDNS by properly starting mDNS on netif
   ...
```
But for me it is still not working, tried with vanilla MP 1.26 in Pico W.

**Reply by ondrej1024** [reply, marked as answer]
I just retried today. Not sure what I changed but it is actually working as expected.
Just established a Wifi connection to my local network with the Pico W and now I can successfully ping `picow.local`.
The default hostname can be changed with `network.hostname("myname")` before bringing up the connection on the Pico W.

**Reply by andypiper** [reply]
> I just retried today.

Was this on the same MP build?

**Reply by ondrej1024** [reply]
Yes, MP 1.26, didn't change it
